File: cli/golem-cli/templates/rust/human-in-the-loop/component-dir/src/workflow_agent.rs
use crate::decision::{Decision, WorkflowId};
use crate::human_agent::HumanAgentClient;
use golem_rust::{
    agent_definition, agent_implementation, await_promise_json, create_promise, description,
    endpoint, prompt,
};
use uuid::Uuid;

#[agent_definition(mount = "/workflows", phantom_agent = true)]
pub trait WorkflowAgent {
    fn new() -> Self;

    #[prompt("Start approval process")]
    #[description("Starts a workflow that requires human approval before continuing")]
    async fn start(&mut self) -> String;
}

struct WorkflowAgentImpl {
    id: WorkflowId,
}

#[agent_implementation]
impl WorkflowAgent for WorkflowAgentImpl {
    fn new() -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
        }
    }

    #[endpoint(post = "/")]
    async fn start(&mut self) -> String {
        // 1. Create a promise that represents waiting for human input
        let approval_promise_id = create_promise();

        // Normally you would send this ID to some UI, email, etc.
        // For demo purposes, we'll just tell you where to send the approval
        let mut approver = HumanAgentClient::get("bob".to_string());
        approver
            .request_approval(self.id.clone(), approval_promise_id.clone())
            .await;

        // 2. Pause here until the promise is completed
        let result: Decision = await_promise_json(&approval_promise_id)
            .await
            .expect("Invalid promise payload");

        // 3. Continue based on human input
        if result == Decision::Approved {
            format!("Workflow {} was approved ✅", self.id)
        } else {
            format!("Workflow {} was rejected ❌", self.id)
        }
    }
}

File: cli/golem-cli/templates/ts/human-in-the-loop/component-dir/src/workflow-agent.ts
import { z } from 'zod';
import { defineAgent, method, http, clientFor, createPromise, awaitPromise } from '@golemcloud/golem-ts-sdk';
import { HumanAgent, encodePromiseId } from './human-agent.js';

// A typed RPC client factory for the remote HumanAgent (wasm-RPC under the hood).
const humanClient = clientFor(HumanAgent);

// The workflow side of the loop: it creates a promise, hands it to a human for
// approval, then PAUSES until the promise is completed — the classic
// human-in-the-loop pattern. Each workflow instance gets its own generated id.
export const WorkflowAgent = defineAgent({
  name: 'WorkflowAgent',
  id: { name: z.string() },
  http: http.mount('/workflows/{name}'),
  methods: {
    start: method({ input: { approver: z.string() }, returns: z.string(), http: http.post('/start') }),
  },
});

export const WorkflowAgentImpl = WorkflowAgent.implement({
  init: () => ({ workflowId: crypto.randomUUID() as string }),
  methods: {
    async start({ approver }) {
      // 1. Create a promise that represents waiting for human input.
      const approvalPromiseId = createPromise();

      // 2. Register the pending approval with the human (remote agent call).
      //    Normally you would surface this in a UI, email, etc.
      await humanClient({ username: approver }).requestApproval({
        workflowId: this.workflowId,
        promiseId: encodePromiseId(approvalPromiseId),
      });

      // 3. Pause here until the promise is completed by the human.
      const resultBytes = await awaitPromise(approvalPromiseId);
      const decision = new TextDecoder().decode(resultBytes);

      // 4. Continue based on the human decision.
      return decision === 'approved'
        ? `Workflow ${this.workflowId} was approved`
        : `Workflow ${this.workflowId} was rejected`;
    },
  },
});
